src/codegen/riscv64/assembler-riscv64.cc

```diff
@@ -850,6 +850,23 @@ void Assembler::GenInstrJ(Opcode opcode, Register rd, int32_t imm21) {
   emit(instr);
 }
 
+void Assembler::GenInstrCR(uint8_t funct4, Opcode opcode, 
+                          Register rd, Register rs2) {
+  DCHECK(is_uint4(funct4) && rd.is_valid() && rs2.is_valid());
+  ShortInstr instr = opcode | (rs2.code() << kRvcRs2Shift) | (rd.code() << kRvcRdShift) |
+                   (funct4 << kRvcFunct4Shift);
+  emit(instr);
+}
+
+void Assembler::GenInstrCI(uint8_t funct3, Opcode opcode, Register rd,
+                           int16_t imm6) {
+  DCHECK(is_uint3(funct3) && rd.is_valid() &&
+         (is_uint6(imm6) || is_int6(imm6)));
```

Comment:
Let's separate GenInstrCI into two helpers. One expects `is_int6(imm6)` (say `GenInstrCI()`), the other expects `is_uint6(imm6)` (say `GenInstrCIU()`) for shift instructions. This way we can check for sure the immediate fields are in range.